File: leetcode/medium/first_missing_positive.java
// https://leetcode.com/problems/first-missing-positive/

class Solution {
    public int firstMissingPositive(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();
        
        for(int i: nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        
        for(int i = 1; i <= (Math.pow(2, 31) - 1); i++) {
            if(!map.containsKey(i)) return i;
        }
        return 1;
    }
}

File: binarysearch.com/easy/questions/array/first_missing_positive.cpp
// https://binarysearch.com/problems/First-Missing-Positive

int solve(vector<int>& nums1) {
    if(nums1.size()==0) return 1;
    if(nums1.size()==1 && nums1[0]==1) return 2;
    else if(nums1.size()==1) return 1;
    set<int> nums;
    for(auto i: nums1) {
        if(i>0) {
            nums.insert(i);
        }
    }
    if(nums.size()==0) return 1;
    auto num=nums.begin();
    int previous = *num;
    if(previous>1) return 1;
    num++;
    for(;num!=nums.end();num++) {
        if(previous+1 != *num) {
            cout<<"in here"<<previous<<" "<<*num;
            break;
        }
        previous = *num;
    }
    if(previous<=0) return 1;
    else return previous+1;
}